Short-term fix to preserve NJT metadata cache in torch.compile

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122836

Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

jbschlosser committed:
Short-term fix to preserve NJT metadata cache in torch.compile

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55448636).

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -2799,8 +2799,14 @@
+
 - name: _nested_get_values(Tensor(a) self) -> Tensor(a)
-  self: _nested_view_from_jagged(grad, at::_nested_get_offsets(self), at::_nested_get_jagged_dummy(self), at::_nested_get_lengths(self), at::_nested_get_ragged_idx(self))
+  self: _nested_view_from_jagged_with_cache(grad, at::_nested_get_offsets(self), at::_nested_get_jagged_dummy(self), at::_nested_get_lengths(self), at::_nested_get_ragged_idx(self), at::_nested_get_min_seqlen(self), at::_nested_get_max_seqlen(self))
```
Would be nice to have a test that repro's @YuqingJ's original error. Maybe compile + using this function + backward would do the trick?

jbschlosser replied:
yep agreed, i did add a test that fails propagating the cache, but we could go a step further and trigger the new error from Brian's recent PR. I'll look into it

jbschlosser replied:
Okay looks like adding a `backward()` call does the trick; I was able to trigger the error

YuqingJ commented on torch/nested/_internal/ops.py:
```diff
@@ -1076,6 +1076,38 @@ def _nested_view_from_jagged_default(func, *args, **kwargs):
+    )
+    ragged_idx = new_kwargs["ragged_idx"]
+    min_seqlen = new_kwargs["min_seqlen"]
+    max_seqlen = new_kwargs["min_seqlen"]
```
typo

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6200,6 +6200,16 @@
   device_check: NoCheck
   dispatch: {}
 
+- func: _nested_get_min_seqlen(Tensor self) -> int
```
I think these need to be symbolic to work with dynamic shapes. TBD exploring this

jbschlosser replied:
This needs more thought / design work to fully cover dynamic shapes. @YuqingJ mentioned to me that a static max length is being used for now, so I expect the patch to work for FIRST.

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

clee2000 commented:
@clee2000 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55448636).

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser commented:
@jbschlosser has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55448636).

jbschlosser committed:
Update

[ghstack-poisoned]

jbschlosser committed:
Update

[ghstack-poisoned]

soulitzer commented on torch/nested/_internal/ops.py:
```diff
@@ -1112,6 +1125,26 @@ def _nested_get_ragged_idx(func, *args, **kwargs):
+    inp = new_kwargs.pop("input")
+    return inp._metadata_cache.get("max_seqlen", None)
+
+
```
Curious if it is possible to do this via an allow_in_graph'd function instead? 

Both would show up in the dynamic graph, and then aot autograd would end up desugaring them into accessing shape on inner tensor.

jbschlosser replied:
If you're referring to the `_nested_get_{min,max}_seqlen()` ops, we also need access to them from C++ (for functional inverses / derivative formula), so I don't think an `allow_in_graph()`d function is enough.

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -4339,6 +4353,115 @@ def forward(self, query, value, offsets):
+            return torch.ones_like(nt) * nt.max_seqlen()
+
+        for dynamic in [False, True]:
+            self.assertFalse(_recompiles_for_inputs(f, (nt,), (nt2,), dynamic=dynamic))
```
jfyi there's `with unittest.mock.patch("torch._dynamo.config.error_on_recompile", True):` though `_recompiles_for_inputs` seems generally useful if you also want to make sure cases where recompiles=True

jbschlosser replied:
I lifted `_recompiles_for_inputs()` from some other test file, which isn't great :p

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -26,6 +26,11 @@ def _get_sdpa_extreme_seqlen(func, tensor):
 
+def _store_val_in_tensor(val) -> torch.Tensor:
+    # hack to get dynamic shapes support: store in a (val, 0) shaped tensor
+    return torch.zeros(val, 0)
```
nit: have both load and store in functions or not for symmetry?

davidberard98 commented on test/test_nestedtensor.py:
```diff
@@ -4339,6 +4353,115 @@ def forward(self, query, value, offsets):
         self.assertTrue(torch.allclose(attn_output_eager, attn_output))
         self.assertTrue(torch.allclose(value_grad, value.grad))
 
+    @dtypes(torch.float32)
```
do you know if this will work with a max_seqlen manually assigned in the graph? (and if so, can you add a test / or add as a followup)

jbschlosser replied:
good Q! this works with manual assignment as done in FIRST (i.e. via `convert_jagged_to_nested_tensor() -> ViewNestedFromBuffer` with the metadata cache info). Also see `test_dummy_mha_with_nt` which tries to match as closely as possible to what the FIRST model does.

We should arguably support and test manual max seq len assignment for the new API. I'll address this or add a TODO

jbschlosser replied:
Update: I think I'll boot handling manual assignment outside of the FIRST use case until we can have public `@property` getters / setters for min / max sequence length that work with PT2 (this is in progress)

jbschlosser committed:
Update

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

soulitzer commented on test/test_nestedtensor.py:
```diff
@@ -4347,6 +4361,138 @@ def forward(self, query, value, offsets):
+            torch.randn(11, 5),
+        ], layout=torch.jagged)
+
+        # TODO: demonstrate usage of new API with manually set seq lengths
```
Is this TODO planned for this PR?

jbschlosser replied:
Yeah, I'm working on this today :) might boot it still, TBD

jbschlosser replied:
Okay, I think I'm going to boot it. In my head, the ideal state is to have public `@property` accessors for both getting and setting min / max sequence length. There's some work being done right now for better `@property` tracing, so I think we should wait on that.

soulitzer reviewed (approved):
Thanks, looks great!

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot rebase

jbschlosser committed:
Update

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

davidberard98 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -6185,12 +6185,12 @@
   autogen: _nested_view_from_buffer_copy.out
 
-- func: _nested_view_from_jagged(Tensor(a) self, Tensor offsets, Tensor dummy, Tensor? lengths=None, int ragged_idx=1) -> Tensor(a)
+- func: _nested_view_from_jagged(Tensor(a) self, Tensor offsets, Tensor dummy, Tensor? lengths=None, int ragged_idx=1, Tensor? min_seqlen=None, Tensor? max_seqlen=None) -> Tensor(a)
```
quick question, do we want to eventually make max/min_seqlen available from the torch.nested.nested_tensor_from_jagged API?

jbschlosser replied:
We definitely need a way for the user to provide these manually. Adding them to the `nested_tensor_from_jagged()` API is one way to do this. One additional way / alternative is to provide public setter `@properties` for min / max sequence length. Some better Dynamo tracing support for `@properties` landed recently so this may be feasible now.

I was originally thinking we'd only allow manual setting through setter properties, but I could probably be convinced that the `nested_tensor_from_jagged()` API should accept these too. Opinions on this?

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@davidberard98 did you find that there are necessary internal changes to land this diff or can we just unlink and re-import?

davidberard98 commented:
@pytorchbot merge

jbschlosser commented:
> @jbschlosser I unlinked it. Main issue was just the debug builds failing

@davidberard98 thanks! Fix for debug build stuff was landed in #128057

jbschlosser commented:
@pytorchbot merge -i

huydhn commented:
@jbschlosser The test is failing ROCm https://hud.pytorch.org/pytorch/pytorch/commit/2a41fc03903de63270d325bd1886a50faf32d7e4#26325533680, plz help take a look

jbschlosser commented:
@huydhn can we just forward fix and skip the test in ROCm?

huydhn commented:
> @huydhn can we just forward fix and skip the test in ROCm?

Yes, plz go ahead with the fix, I can stamp it if you need

huydhn commented:
@jbschlosser I have also just noticed another periodic failure coming out of this PR https://hud.pytorch.org/pytorch/pytorch/commit/2a41fc03903de63270d325bd1886a50faf32d7e4#26340619959.  It's a CUDA memory leak failure (we only run memory leak check periodically) and your PR is the only one touching  `test_nestedtensor.py` during the bisect window:

```
FAILED [1.0717s] test_nestedtensor.py::TestNestedTensorSubclassCUDA::test_dummy_mha_with_nt_cuda - RuntimeError: CUDA driver API confirmed a leak in __main__.TestNestedTensorSubclassCUDA.test_dummy_mha_with_nt_cuda! Caching allocator allocated memory was 8192 and is now reported as 12288 on device 0. CUDA driver allocated memory was 356909056 and is now 359006208.
2024-06-17T23:21:21.1774328Z 
```

Any thoughts?

davidberard98 commented:
btw, I noticed a week or two ago that https://github.com/pytorch/pytorch/issues/125058 was failing similarly, but I wasn't able to repro

cc @drisspg , in case you know whether any of the attention implementations would have this issue?

jbschlosser commented:
@pytorchbot revert

jbschlosser commented:
@pytorchbot revert -m "internal test failures with DEBUG=1 asserts"

jbschlosser commented:
@pytorchbot revert -m "internal test failures with DEBUG=1 asserts" -c nosignal

jbschlosser commented:
Plan to reland with:
* ROCm test skipping (#128891)
* DEBUG=1 asserts disabled for `_nested_get_{min,max}_seqlen()` (#129014)
* memory leak fix? or at least investigation into this

jbschlosser commented:
@pytorchbot revert -m "internal test failures with DEBUG=1 asserts" -c nosignal

jbschlosser commented:
@pytorchbot revert -m "internal test failures with DEBUG=1 asserts" -c nosignal

huydhn commented:
Just FYI, we have a small tool to detect dependencies here https://github.com/pytorch/test-infra/actions/workflows/pr-dependencies-check.yml.  I run the job with this PR and get the following list of conflicts https://github.com/pytorch/test-infra/actions/runs/9573995289/attempts/1#summary-26396464107.  This is an easy case because they are all in your stack, but the tool might be useful to find conflicts from elsewhere.

jbschlosser commented:
> Just FYI, we have a small tool to detect dependencies here https://github.com/pytorch/test-infra/actions/workflows/pr-dependencies-check.yml. I run the job with this PR and get the following list of conflicts https://github.com/pytorch/test-infra/actions/runs/9573995289/attempts/1#summary-26396464107. This is an easy case because they are all in your stack, but the tool might be useful to find conflicts from elsewhere (if it happens)

wow that's super useful! thanks for the link :)

jbschlosser commented:
Confirmed that the memory leak issues are pre-existing and not introduced by this PR. They need more investigation, but I'm not even convinced these are NJT-related. For example, I saw a memory leak from running this minimal, non-NJT example with `PYTORCH_TEST_CUDA_MEM_LEAK_CHECK=1`:
```python
def test_memory_leak(self, device):
    value = torch.randn(6, 4, requires_grad=True, device=device)
    m = torch.nn.Linear(4, 6, device=device)
    symbolic_traced: torch.fx.GraphModule = torch.fx.symbolic_trace(m)
    m = torch.compile(symbolic_traced)
    m(value)
```

I'm going to go ahead and reland this with the ROCm / DEBUG=1 assert fixes added.

jbschlosser committed:
Update on "Short-term fix to preserve NJT metadata cache in torch.compile"


Idea: close over min / max sequence length in the main NJT view func (`_nested_view_from_jagged`) so that view replay during fake-ification propagates these correctly in torch.compile.

For dynamic shapes support for min / max sequence length, this PR uses a hack that stores the values in `(val, 0)` shaped tensors.

**NB: This PR changes SDPA to operate on real views instead of using `buffer_from_jagged()` / `ViewNestedFromBuffer`, which may impact the internal FIRST model. That is, it undoes the partial revert from #123215 alongside a fix to the problem that required the partial revert. We need to verify that there are no regressions there before landing.**

Differential Revision: [D55448636](https://our.internmc.facebook.com/intern/diff/D55448636)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i
